File: v1/src/main/java/com/api/v1/services/CreditCardRegistrationServiceImpl.java
package com.api.v1.services;

import com.api.v1.builders.CreditCardBuilder;
import com.api.v1.domain.CreditCard;
import com.api.v1.domain.CreditCardRepository;
import com.api.v1.dtos.CreditCardRegistrationRequestDto;
import com.api.v1.dtos.CreditCardResponseDto;
import com.api.v1.exceptions.DuplicatedCardNumberException;
import com.api.v1.mappers.CreditCardResponseMapper;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class CreditCardRegistrationServiceImpl implements CreditCardRegistrationService {

    @Autowired
    private CreditCardRepository repository;

    @Override
    public Mono<CreditCardResponseDto> register(@Valid CreditCardRegistrationRequestDto requestDto) {
        return repository
                .findAll()
                .filter(e -> e.getCardNumber().equals(requestDto.cardNumber()))
                .hasElements()
                .flatMap(exists -> {
                    if (exists) return Mono.error(new DuplicatedCardNumberException(requestDto.cardNumber()));
                    return Mono.defer(() -> {
                        CreditCard creditCard = CreditCardBuilder.create().fromDto(requestDto).build();
                        Mono<CreditCard> savedCreditCard = repository.save(creditCard);
                       return savedCreditCard.flatMap(CreditCardResponseMapper::mapToMono);
                    });
                });
    }

}

File: v2/src/main/kotlin/com/api/v2/creditcard/CreditCardRegistrationServiceImpl.kt
package com.api.v2.creditcard

import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class CreditCardRegistrationServiceImpl: CreditCardRegistrationService {

    @Autowired
    lateinit var creditCardRepository: CreditCardRepository

    override suspend fun register(creditCard: @Valid CreditCard): CreditCardResponseDTO {
        return withContext(Dispatchers.IO) {
            val isThereADuplicatedCardNumber = creditCardRepository
                .findAll()
                .filter { e -> e.cardNumber == creditCard.cardNumber }
                .firstOrNull() != null
            if (isThereADuplicatedCardNumber) {
                throw DuplicatedCardNumberException(creditCard.cardNumber)
            }
            val savedCredit = creditCardRepository.save(creditCard)
            CreditCardResponseMapper.map(savedCredit)
        }
    }

}